src/StackExchange.Redis/RedisDatabase.cs

```diff
@@ -3172,32 +3208,99 @@ private Message GetSortedSetCombineAndStoreCommandMessage(SetOperation operation
             {
                 SetOperation.Intersect => RedisCommand.ZINTERSTORE,
                 SetOperation.Union => RedisCommand.ZUNIONSTORE,
+                SetOperation.Difference =>
+                    // Difference can't have weights or aggregation
+                    weights != null || aggregate != Aggregate.Sum
+                    ? throw new ArgumentException("ZDIFFSTORE cannot be used with weights or aggregation.")
+                    : RedisCommand.ZDIFFSTORE,
                 _ => throw new ArgumentOutOfRangeException(nameof(operation)),
             };
-            if (keys == null) throw new ArgumentNullException(nameof(keys));
+            if (keys == null)
+            {
+                throw new ArgumentNullException(nameof(keys));
+            }
+            if (weights != null && keys.Length != weights.Length)
+            {
+                throw new ArgumentException("Keys and weights should have the same number of elements.", nameof(weights));
+            }
 
-            List<RedisValue>? values = null;
-            if (weights != null && weights.Length != 0)
+            RedisValue[] values = RedisValue.EmptyArray;
+
+            var argsLength = (weights?.Length > 0 ? 1 + weights.Length : 0) + (aggregate != Aggregate.Sum ? 2 : 0);
+            if (argsLength > 0)
             {
-                (values ??= new List<RedisValue>()).Add(RedisLiterals.WEIGHTS);
+                values = new RedisValue[argsLength];
+                AddWeightsAggregationAndScore(values, weights, aggregate);
+            }
+            return new SortedSetCombineAndStoreCommandMessage(Database, flags, command, destination, keys, values);
+        }
+
+        private Message GetSortedSetCombineCommandMessage(SetOperation operation, RedisKey[] keys, double[]? weights, Aggregate aggregate, bool withScores, CommandFlags flags)
+        {
+            var command = operation switch
+            {
+                SetOperation.Intersect => RedisCommand.ZINTER,
+                SetOperation.Union => RedisCommand.ZUNION,
+                SetOperation.Difference =>
+                    // Difference can't have weights or aggregation
+                    weights != null || aggregate != Aggregate.Sum
+                    ? throw new ArgumentException("ZDIFF cannot be used with weights or aggregation.")
+                    : RedisCommand.ZDIFF,
+                _ => throw new ArgumentOutOfRangeException(nameof(operation)),
+            };
+            if (keys == null)
+            {
+                throw new ArgumentNullException(nameof(keys));
+            }
+            if (weights != null && keys.Length != weights.Length)
+            {
+                throw new ArgumentException("Keys and weights should have the same number of elements.", nameof(weights));
+            }
+
+            var i = 0;
+            var values = new RedisValue[1 + keys.Length +
```

Comment:
How much more expensive would it be to initialize a List with this capacity and using `Add`/`AddRange` as opposed to an array and then having to track `i`, this is a bit. . . feels like we're reinventing the wheel here? `List<T>(int)` is effectively just doing the same initialization, since it's initialized with the correct capacity it shouldn't over allocate nor have to dynamically reallocate.

Reply:
Ahh I gotcha. [TIL](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Collections/Generic/List.cs#L1028)